### DISCUSSION TITLE
(T2)(new T1()) type casting puzzle

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#62378] [Q&A]
If I cast T1 to T2, which T1 is derived from, now the object is in a narrow-er type, I am aware that the actual type is still T1, you are just pretending it as T2.
My question is :if you store this variable as T2, after a while will this object's data that's not visible right now in a narrow-er type be deleted? or it will remain T1 forever and you can get your data back (if there's any inside T1 thatsn't from T2) by casting back to T1

**Comment by gfoidl** [comment]
>  or it will remain T1 forever

That's the case.
You created a T1, so it is a T1. 
With e.g. `T2 obj = (T2)(new T1())` the variable `obj` just holds the reference to object of type `T1` created on the heap. The machine executing your code doesn't care which type `obj` is. This is "only" for the language and the compiler to let have you type-safeness. I.e. with `T2 obj` you are allowed to perform operations on `obj` that are declared on `T2`.

As `obj` is actually a `T1` you can safely cast it back, and access further member declared on `T1`.

**Reply by Xyncgas** [reply]
It's just a bit counter instinct from c++ point of view, if I casted a pointer of an object to a narrow-er pointer, how does the garbage collect even know the actual size of the object. well I guess it's possible, maybe object is keeping track of it.

**Reply by Xyncgas** [reply]
dotnet is smart

**Reply by svick** [reply]
@Xyncgas From a C++ point of view, similar code would still work, if you defined a `virtual` destructor in the base class. The difference is that in C#, it always works.

**Reply by gfoidl** [reply]
@Xyncgas also note that from a .NET point of view -- especially from the runtime part -- each object has a header, where additional info is stored like actual type, info for GC, etc. 
For more info see e.g. [Managed object internals, Part 1. The layout](https://devblogs.microsoft.com/premier-developer/managed-object-internals-part-1-layout/).

**Comment by huoyaoyuan** [comment]
You are confusing with objects and references.

With C++/CLI syntax, you are not casting `T1` to `T2`. You are casting `T1^` to `T2^` instead, which are similar to `T1*` and `T2*`, but tracked by GC.
